### PR TITLE
Delphi update rest api 2 Flask side

### PR DESCRIPTION
**Pull request by adarshp**

### CODE REVIEW

**Comment by manujinda** [comment]
As I understand the immediate response of the experiments route (createCausemosExperiment) is (lines 398 - 406):

    return jsonify(
        {
            "modelId": modelID,
            "experimentId": experiment_id,
            "experimentType": experiment_type,
            "status": "in progress",
            "results": executor.futures._state(experiment_id),
        }
    )

According to the API specification, we have to return:

{
	experimentId: <uuid>
}

for any type of experiment.

**Comment by manujinda** [comment]
In the route getExperimentResults, we have hard coded "experimentType": "PROJECTION", it would be ok for now since PROJECTION is the only experiment we are running.

Ideally we should have a way to query by experiment id and get the type.

**Comment by manujinda** [comment]
We are still missing the response for a failed experiment:

{
    modelId: <uuid>,
    experimentId: <uuid>,
    experimentType: <str>,
    status: "failed",
    results: {}
}

At the moment, run_causemose_projection_experiment() returns an empty CausemosProjectionExperimentResult structure to signal an error. An exception could be more ideal. Idid not want to add that complexity amid this rush. I thought that the current hack would work for now.

The commented lines you deleted:

        #if len(causemos_experiment_result[3]) < numTimesteps:
        #    result.status = "failed"
        #    result.results = {}
        #    db.session.add(result)
        #    db.session.commit()
        #    return

Are Aishwarya's attempt to follow my instructions. My idea was that if an experiment fails, it should create the failed experiment json and store in the database. When asked for results that will be returned.

**Comment by manujinda** [comment]
I cloned your branch, compiled it and ran the test_rest_api.py

For me create experiment fails:

――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――― test_createExperiment ――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――

client = <FlaskClient <Flask 'delphi.apps.rest_api'>>

    def test_createExperiment(client):
        with open(
            "tests/data/delphi/causemos_experiments_projection_input.json",
            encoding="utf-8",
        ) as f:
            data = json.load(f)
        model_id = "XYZ"
        rv = client.post(f"/delphi/models/{model_id}/experiments", json=data)
        experiment_id = rv.get_json()["experimentId"]
        status = rv.get_json()["status"]
        while status != "completed":
            time.sleep(1)
>           rv = client.get(f"/delphi/models/{model_id}/experiments/{experiment_id}")

tests/wm/test_rest_api.py:61: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
../../../../venv/ivilab/lib/python3.8/site-packages/werkzeug/test.py:1006: in get
    return self.open(*args, **kw)
../../../../venv/ivilab/lib/python3.8/site-packages/flask/testing.py:222: in open
    return Client.open(
../../../../venv/ivilab/lib/python3.8/site-packages/werkzeug/test.py:970: in open
    response = self.run_wsgi_app(environ.copy(), buffered=buffered)
../../../../venv/ivilab/lib/python3.8/site-packages/werkzeug/test.py:861: in run_wsgi_app
    rv = run_wsgi_app(self.application, environ, buffered=buffered)
../../../../venv/ivilab/lib/python3.8/site-packages/werkzeug/test.py:1096: in run_wsgi_app
    app_rv = app(environ, start_response)
../../../../venv/ivilab/lib/python3.8/site-packages/flask/app.py:2464: in __call__
    return self.wsgi_app(environ, start_response)
../../../../venv/ivilab/lib/python3.8/site-packages/flask/app.py:2450: in wsgi_app
    response = self.handle_exception(e)
../../../../venv/ivilab/lib/python3.8/site-packages/flask/app.py:1867: in handle_exception
    reraise(exc_type, exc_value, tb)
../../../../venv/ivilab/lib/python3.8/site-packages/flask/_compat.py:39: in reraise
    raise value
../../../../venv/ivilab/lib/python3.8/site-packages/flask/app.py:2447: in wsgi_app
    response = self.full_dispatch_request()
../../../../venv/ivilab/lib/python3.8/site-packages/flask/app.py:1952: in full_dispatch_request
    rv = self.handle_user_exception(e)
../../../../venv/ivilab/lib/python3.8/site-packages/flask/app.py:1821: in handle_user_exception
    reraise(exc_type, exc_value, tb)
../../../../venv/ivilab/lib/python3.8/site-packages/flask/_compat.py:39: in reraise
    raise value
../../../../venv/ivilab/lib/python3.8/site-packages/flask/app.py:1950: in full_dispatch_request
    rv = self.dispatch_request()
../../../../venv/ivilab/lib/python3.8/site-packages/flask/app.py:1936: in dispatch_request
    return self.view_functions[rule.endpoint](**req.view_args)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

modelID = 'XYZ', experimentID = '34a53633-ab66-44df-b3ac-d27688a74125'

    @bp.route(
        "/delphi/models/<string:modelID>/experiments/<string:experimentID>",
        methods=["GET"],
    )
    def getExperimentResults(modelID: str, experimentID: str):
        """ Fetch experiment results"""
        if not executor.futures.done(experimentID):
            return jsonify(
                {
                    "experimentId": experimentID,
                    "status": executor.futures._state(experimentID),
                }
            )
        else:
            experimentResult = CauseMosForwardProjectionResult.query.filter_by(
                id=experimentID
            ).first()
            return jsonify(
                {
                    "experimentId": experimentID,
>                   "results": experimentResult.deserialize()["results"],
                    "status": "COMPLETE",
                }
            )
E           AttributeError: 'NoneType' object has no attribute 'deserialize'

../delphi/delphi/apps/rest_api/api.py:421: AttributeError

 tests/wm/test_rest_api.py ⨯s                                                                                                                                                                                          100% ██████████
====================================================================================================== short test summary info =======================================================================================================
FAILED tests/wm/test_rest_api.py::test_createExperiment - AttributeError: 'NoneType' object has no attribute 'deserialize'

Results (14.03s):
       1 passed
       1 failed
         - tests/wm/test_rest_api.py:49 test_createExperiment
       1 skipped

shell returned 1

**Comment by adarshp** [comment]
Thanks @manujinda  - I will try to address these later this afternoon. Probably can't get to it until 4 pm though.

File: delphi/apps/rest_api/api.py [hunk 1]
```diff
@@ -404,112 +424,18 @@ def runExperiment():
...
+    ).first()
 
+    response = {
+        "modelId": modelID,
+        "experimentId": experimentID,
+        "experimentType": result.experimentType,
+        "status": result.status,
+        "results": result.results
+    }
+    return jsonify(response)
```
[the remark this answers is not in the document]
Comment (manujinda) [reply delphi/apps/rest_api/api.py:438]: Oh I see that you are achieving this in a different way. You are writing the in-progress message. to the database. It seems more cleaver. 

I think we are good to go.

I'll merge the branches and do some more cleaning and do a pull request.

**Review by manujinda — APPROVED** [review]

**Comment by manujinda** [comment]
I think we are good to go.

Thank you very much Adarsh.

**Pull request closed without merging**